File: packages/sdk-python/veto/capsule/rfc3339.py
"""Strict RFC 3339 / ISO 8601 parser — Python mirror of rfc3339.ts.

Matches TypeScript behavior component-for-component so the two SDKs agree on
every timestamp. Rejects impossible dates (2026-02-31), out-of-range offsets
(+24:00, +00:60), and precision a human couldn't express portably
(fractional >6 digits).
"""

from __future__ import annotations

import calendar
import datetime as _dt
import re
from dataclasses import dataclass

_RE = re.compile(
    r"^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(?:\.(\d{1,6}))?(Z|[+-]\d{2}:\d{2})$"
)


class Rfc3339ParseError(ValueError):
    """Raised when a timestamp string fails strict RFC 3339 parsing."""


@dataclass
class ParsedRfc3339:
    epoch_ms: int
    canonical: str
    offset: str


def parse_rfc3339_strict(value: str) -> ParsedRfc3339:
    if not isinstance(value, str):
        raise Rfc3339ParseError(
            f"timestamp must be a string; got {type(value).__name__}"
        )
    m = _RE.match(value)
    if not m:
        raise Rfc3339ParseError(
            f"timestamp must be RFC 3339 (YYYY-MM-DDTHH:MM:SS[.ffffff](Z|\u00b1HH:MM)); got {value!r}"
        )
    year_s, month_s, day_s, hour_s, min_s, sec_s, frac_s, offset = m.groups()
    year = int(year_s)
    month = int(month_s)
    day = int(day_s)
    hour = int(hour_s)
    minute = int(min_s)
    second = int(sec_s)

    if year < 1:
        raise Rfc3339ParseError(f"year must be >= 0001; got {year_s}")
    if not 1 <= month <= 12:
        raise Rfc3339ParseError(f"month must be 01..12; got {month_s}")
    max_day = calendar.monthrange(year, month)[1]
    if not 1 <= day <= max_day:
        raise Rfc3339ParseError(f"day {day_s} out of range for {year_s}-{month_s}")
    if hour > 23:
        raise Rfc3339ParseError(f"hour must be 00..23; got {hour_s}")
    if minute > 59:
        raise Rfc3339ParseError(f"minute must be 00..59; got {min_s}")
    if second > 60:
        raise Rfc3339ParseError(f"second must be 00..60; got {sec_s}")
    if second == 60:
        # RFC 3339 §5.6: leap seconds occur ONLY at 23:59:60 UTC on the last
        # day of a UTC month. Anything else silently collapsing to :59 makes
        # two semantically-distinct payloads hash-equal after canonicalization.
        if hour != 23 or minute != 59:
            raise Rfc3339ParseError(
                f"leap second :60 is only legal at 23:59:60; got {hour_s}:{min_s}:{sec_s}"
            )
        second = 59  # clamp ONLY at 23:59:60

    offset_minutes = 0
    if offset != "Z":
        sign = 1 if offset.startswith("+") else -1
        o_hour = int(offset[1:3])
        o_min = int(offset[4:6])
        if o_hour > 23:
            raise Rfc3339ParseError(f"offset hour must be 00..23; got {offset}")
        if o_min > 59:
            raise Rfc3339ParseError(f"offset minute must be 00..59; got {offset}")
        offset_minutes = sign * (o_hour * 60 + o_min)

    utc_naive = _dt.datetime(year, month, day, hour, minute, second, tzinfo=_dt.timezone.utc)
    # Use floor (not round) so TS and Python produce the same epoch_ms for
    # every fractional input. Math.floor vs Python round() caused a 1ms
    # drift at boundaries like .123500 — flagged by adversarial codex pass.
    import math as _math
    frac_ms = _math.floor(float("0." + frac_s) * 1000) if frac_s else 0
    epoch_ms = int(utc_naive.timestamp() * 1000) + frac_ms - offset_minutes * 60 * 1000

    # Round-trip canonical UTC form with no trailing-zero fraction.
    dt_utc = _dt.datetime.fromtimestamp(epoch_ms / 1000, tz=_dt.timezone.utc)
    micro_ms = dt_utc.microsecond // 1000
    frac_part = f".{micro_ms:03d}" if micro_ms else ""
    canonical = (
        f"{dt_utc.year:04d}-{dt_utc.month:02d}-{dt_utc.day:02d}"
        f"T{dt_utc.hour:02d}:{dt_utc.minute:02d}:{dt_utc.second:02d}"
        f"{frac_part}Z"
    )

    return ParsedRfc3339(epoch_ms=epoch_ms, canonical=canonical, offset=offset)


def is_valid_rfc3339(value: object) -> bool:
    if not isinstance(value, str):
        return False
    try:
        parse_rfc3339_strict(value)
        return True
    except Rfc3339ParseError:
        return False

File: packages/spend-capsule-protocol/src/rfc3339.ts
// Strict RFC 3339 / ISO 8601 date-time parser. We roll our own because
// `new Date(str)` in TS silently normalizes impossible dates like
// "2026-02-31T00:00:00Z" to "2026-03-03..." (month/day rollover), and Python
// 3.11's datetime.fromisoformat rejects them, producing split-brain
// acceptance across the two SDKs.
//
// What this enforces:
//   - 4-digit year (0001–9999)
//   - month 1..12, day 1..daysInMonth(year, month)
//   - hour 0..23, minute 0..59, second 0..60 (leap second allowed but
//     we round it to :59 for Date interop), no 24:00
//   - fractional precision capped to 6 digits (microsecond; matches Python
//     stdlib). 7+ rejected so TS/Python can never disagree.
//   - offset must be "Z" or [+-]HH:MM with HH 0..23, MM 0..59.
//     "+24:00" and "+00:60" both rejected.

const RFC3339_RE =
  /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(?:\.(\d{1,6}))?(Z|[+-]\d{2}:\d{2})$/;

function isLeapYear(y: number): boolean {
  return (y % 4 === 0 && y % 100 !== 0) || y % 400 === 0;
}

function daysInMonth(y: number, m: number): number {
  switch (m) {
    case 1:
    case 3:
    case 5:
    case 7:
    case 8:
    case 10:
    case 12:
      return 31;
    case 4:
    case 6:
    case 9:
    case 11:
      return 30;
    case 2:
      return isLeapYear(y) ? 29 : 28;
    default:
      return 0;
  }
}

export interface ParsedRfc3339 {
  /** Milliseconds since unix epoch (same as Date.getTime()). */
  epochMs: number;
  /** Round-trip canonical string (UTC, "Z" suffix, no fractional if zero). */
  canonical: string;
  /** Original offset field ("Z" or "+HH:MM" / "-HH:MM"). */
  offset: string;
}

export class Rfc3339ParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "Rfc3339ParseError";
  }
}

export function parseRfc3339Strict(value: string): ParsedRfc3339 {
  if (typeof value !== "string") {
    throw new Rfc3339ParseError(`timestamp must be a string; got ${typeof value}`);
  }
  const m = RFC3339_RE.exec(value);
  if (!m) {
    throw new Rfc3339ParseError(
      `timestamp must be RFC 3339 (YYYY-MM-DDTHH:MM:SS[.ffffff](Z|±HH:MM)); got "${value}"`,
    );
  }
  const [, yearStr, monthStr, dayStr, hourStr, minStr, secStr, fracStr, offset] = m;
  const year = parseInt(yearStr!, 10);
  const month = parseInt(monthStr!, 10);
  const day = parseInt(dayStr!, 10);
  const hour = parseInt(hourStr!, 10);
  const minute = parseInt(minStr!, 10);
  let second = parseInt(secStr!, 10);

  if (year < 1) {
    throw new Rfc3339ParseError(`year must be >= 0001; got ${yearStr}`);
  }
  if (month < 1 || month > 12) {
    throw new Rfc3339ParseError(`month must be 01..12; got ${monthStr}`);
  }
  if (day < 1 || day > daysInMonth(year, month)) {
    throw new Rfc3339ParseError(
      `day ${dayStr} out of range for ${yearStr}-${monthStr}`,
    );
  }
  if (hour > 23) {
    throw new Rfc3339ParseError(`hour must be 00..23; got ${hourStr}`);
  }
  if (minute > 59) {
    throw new Rfc3339ParseError(`minute must be 00..59; got ${minStr}`);
  }
  if (second > 60) {
    throw new Rfc3339ParseError(`second must be 00..60; got ${secStr}`);
  }
  if (second === 60) {
    // RFC 3339 §5.6: leap seconds occur ONLY at 23:59:60 UTC on the last
    // day of a UTC month. Anything else is impossible and must be rejected
    // rather than silently coerced to :59 — otherwise a forged "12:34:60"
    // timestamp verifies identically to "12:34:59", which makes two
    // semantically-distinct payloads hash-equal after canonicalization.
    if (hour !== 23 || minute !== 59) {
      throw new Rfc3339ParseError(
        `leap second :60 is only legal at 23:59:60; got ${hourStr}:${minStr}:${secStr}`,
      );
    }
    second = 59; // clamp ONLY at 23:59:60
  }

  // Offset validation.
  let offsetMinutes = 0;
  if (offset !== "Z") {
    const sign = offset!.startsWith("+") ? 1 : -1;
    const oHour = parseInt(offset!.slice(1, 3), 10);
    const oMin = parseInt(offset!.slice(4, 6), 10);
    if (oHour > 23) {
      throw new Rfc3339ParseError(`offset hour must be 00..23; got ${offset}`);
    }
    if (oMin > 59) {
      throw new Rfc3339ParseError(`offset minute must be 00..59; got ${offset}`);
    }
    offsetMinutes = sign * (oHour * 60 + oMin);
  }

  // Compute epoch.
  const utcMs = Date.UTC(year, month - 1, day, hour, minute, second);
  const fracMs = fracStr ? Math.floor(Number("0." + fracStr) * 1000) : 0;
  const epochMs = utcMs + fracMs - offsetMinutes * 60 * 1000;

  // Round-trip canonical form (UTC, "Z", no trailing zeros).
  const d = new Date(epochMs);
  const pad = (n: number, w: number = 2) => n.toString().padStart(w, "0");
  const fracCanon = d.getUTCMilliseconds();
  const fracPart = fracCanon === 0 ? "" : "." + pad(fracCanon, 3);
  const canonical =
    `${pad(d.getUTCFullYear(), 4)}-${pad(d.getUTCMonth() + 1)}-${pad(d.getUTCDate())}` +
    `T${pad(d.getUTCHours())}:${pad(d.getUTCMinutes())}:${pad(d.getUTCSeconds())}` +
    `${fracPart}Z`;

  return { epochMs, canonical, offset: offset! };
}

export function isValidRfc3339(value: unknown): boolean {
  if (typeof value !== "string") return false;
  try {
    parseRfc3339Strict(value);
    return true;
  } catch {
    return false;
  }
}
